man/grid-updates.1
.TH "GRID\-UPDATES" "1" "September 2012" "User Commands" ""
.SH NAME
.PP
grid\-updates \- Helper Script for Tahoe\-LAFS Nodes
.SH SYNOPSIS
.PP
\f[B]grid\-updates\f[] [\f[I]OPTIONS\f[]] \f[I]ACTIONS\f[]
.SH DESCRIPTION
.PP
\f[I]grid\-updates\f[] is a Python script intended to help keep
Tahoe\-LAFS nodes\[aq] configurations up\-to\-date.
It can retrieve lists of introducers as well as news feeds from the
Tahoe grid.
This is useful for any public grid that relies solely on volunteers.
.SH ACTIONS
.TP
.B \-s, \-\-sync\-introducers
Synchronize your node\[aq]s local list of introducers with the master
list.
.RS
.RE
.TP
.B \-m, \-\-merge\-introducers
Merge your node\[aq]s local introducers list with the
subscription\[aq]s.
.RS
.RE
.TP
.B \-n, \-\-download\-news
Retrieve the news feed.
See the \f[B]NEWS\f[] section below.
.RS
.RE
.TP
.B \-r, \-\-repair
Maintain the health of Tahoe shares listed in a subscription.
.RS
.RE
.TP
.B \-\-patch\-tahoe
Patch the Tahoe\-LAFS web console to display the grid\-updates news feed
in an Iframe.
.RS
.RE
.TP
.B \-\-undo\-patch\-tahoe
Remove the grid\-updates patch to the Tahoe web console and restore its
original version.
.RS
.RE
.TP
.B \-\-make\-news \f[I]file\f[]
Create a \f[I]grid\-updates\f[]\-compatible NEWS.tgz file from a
reStructuredText source file.
.RS
.RE
.TP
.B \-\-check\-version
Check for a new version of this script on the grid.
.RS
.RE
.TP
.B \-\-download\-update
Download a new version of this script from the grid to the specified
directory (implies \f[C]\-\-check\-update\f[]).
.RS
.RE
.SH OPTIONS
.TP
.B \-c \f[I]config file\f[], \-\-config \f[I]config file\f[]
Specify a grid\-updates config file.
See \f[B]CONFIG FILES\f[] section below.
.RS
.RE
.TP
.B \-u \f[I]URL\f[], \-\-node\-url \f[I]URL\f[]
Specify the gateway node\[aq]s URL (default: http://127.0.0.1:3456),
.RS
.RE
.TP
.B \-d \f[I]directory\f[], \-\-node\-directory \f[I]directory\f[]
Specify the node directory (default: \f[I]~/.tahoe\f[]),
.RS
.RE
.TP
.B \-\-format \f[I]tar|deb|zip|exe|py2exe\f[]
Specify in which format to download the update.
Choices are: \[aq]tar\[aq] (unix source archive), \[aq]deb\[aq] (Debian
package), \[aq]zip\[aq] (Windows source archive), \[aq]exe\[aq] (Windows
installer [requires Python]), \[aq]py2exe\[aq] (Windows installer
[doesn\[aq]t require Python]).
.RS
.RE
.TP
.B \-\-list\-uri \f[I]FILE CAP\f[]
Override the default location of the introducers list.
.RS
.RE
.TP
.B \-\-news\-uri \f[I]FILE CAP\f[]
Override the default location of the NEWS.tgz file.
.RS
.RE
.TP
.B \-\-script\-uri \f[I]DIR CAP\f[]
Override the default location of script updates.
.RS
.RE
.TP
.B \-\-repairlist\-uri \f[I]FILE CAP\f[]
Override the default location of the \-\-repair subscription file.
.RS
.RE
.TP
.B \-v
Increase verbosity of output.
.RS
.RE
.TP
.B \-V, \-\-version
Display version information.
.RS
.RE
.TP
.B \-h, \-\-help
Print usage information.
.RS
.RE
.SH CONFIG FILES
.PP
\f[I]grid\-updates\f[] will look for configuration files in
\f[I]X\f[]\f[I]D\f[]\f[I]G\f[]~\f[I]C\f[]~\f[I]O\f[]\f[I]N\f[]\f[I]F\f[]\f[I]I\f[]\f[I]G\f[]~\f[I]H\f[]~\f[I]O\f[]\f[I]M\f[]\f[I]E\f[]/\f[I]g\f[]\f[I]r\f[]\f[I]i\f[]\f[I]d\f[] − \f[I]u\f[]\f[I]p\f[]\f[I]d\f[]\f[I]a\f[]\f[I]t\f[]\f[I]e\f[]\f[I]s\f[]/\f[I]c\f[]\f[I]o\f[]\f[I]n\f[]\f[I]f\f[]\f[I]i\f[]\f[I]g\f[].\f[I]i\f[]\f[I]n\f[]\f[I]i\f[]\f[I]a\f[]\f[I]n\f[]\f[I]d\f[]XDG_CONFIG_DIRS/grid\-updates/config.ini.
.PP
All options must be listed below the section \f[C][OPTIONS]\f[].
.PP
Accepted options are:
.TP
.B tahoe_node_dir = \f[I]directory\f[]
Same as \-\-node\-dir above
.RS
.RE
.TP
.B list_uri = \f[I]FILE CAP\f[]
Same as \-\-list\-uri option above
.RS
.RE
.TP
.B news_uri = \f[I]FILE CAP\f[]
Same as \-\-news\-uri option above
.RS
.RE
.TP
.B script_uri = \f[I]DIR CAP\f[]
Same as \-\-script\-uri option above
.RS
.RE
.TP
.B comrepair_uri = \f[I]FILE CAP\f[]
Same as \-\-comrepair\-uri option above
.RS
.RE
.SH NEWS
.PP
If you choose to download the news feed, \f[I]grid\-updates\f[] will
place a plain text version of it in your node\[aq]s directory and print
it to stdout.
This is intended to be sent by cron mail.
.PP
There is also an HTML version of the news feed that you can view in a
web browser.
You can access it either directly
(http://127.0.0.1:3456/static/NEWS.html) or have it be displayed on your
Tahoe node\[aq]s web console.
You can prepare the console by running
\f[C]grid\-updates\ \-\-patch\-tahoe\f[] once.
.PP
The Atom news feed (http://127.0.0.1:3456/static/NEWS.atom) can be used
by regular feed readers to check for \f[I]grid\-updates\f[] news.
(Please note, however, that you cannot "refresh" the feed with regular
news readers.
These files have to always be fetched by \f[I]grid\-updates\f[] first.)
.SH INFORMATION FOR SUBSCRIPTION MAINTAINERS
.PP
If you want to offer a \f[I]grid\-updates\f[] subscription service, you
will have to provide users with the URIs of your subscription files.
They can be either in the form \f[I]DIR CAP/filename\f[] or the the file
cap itself.
In the latter case you must make sure to create mutable files.
.PP
For the specific requirements of the JSON lists, please see the included
example files in the share directory.
.SH FILES
.IP \[bu] 2
\f[I]~/.tahoe/introducers\f[]
.PD 0
.P
.PD
.IP \[bu] 2
\f[I]~/.tahoe/NEWS\f[]
.PD 0
.P
.PD
.IP \[bu] 2
\f[I]~/.tahoe/public_html/NEWS.html\f[]
.PD 0
.P
.PD
.IP \[bu] 2
\f[I]~/.tahoe/public_html/NEWS.atom\f[]
.PD 0
.P
.PD
.IP \[bu] 2
\f[I]$XDG_CONFIG_HOME/grid\-updates/config\f[] (most commonly ~/.config)
.PD 0
.P
.PD
.IP \[bu] 2
\f[I]$XDG_CONFIG_DIRS/grid\-updates/config\f[] (most commonly /etc/xdg)
.SH BUGS
.TP
.B \f[B]ERROR: Can\[aq]t parse JSON list: No JSON object could be
decoded\f[]
Currently Tahoe doesn\[aq]t return JSON data if it encounters exceptions
(see Trac ticket #1799).
If you see this \f[B]grid\-updates\f[] error, you can rerun your command
in debug mode (\f[C]\-vvv\f[]) to see Tahoe\[aq]s actual response.
.RS
.PP
This error is most likely to occur during \f[I]deep\-check\f[]
operations.
If it does, it probably encountered the \f[I]NotEnoughSharesError\f[]
error, which means that a file was unrecoverable.
You should investigate the problem using Tahoe directly.
.RE
.TP
.B \f[B]ERROR: Could not run one\-check for testfile: HTTP Error 410:
Gone\f[]
This error is related to the one above but happens during
\f[I]one\-check\f[] operations.
If a file is not retrievable (due to not enough remaining shares) Tahoe
responds with HTTP error 410.
.RS
.RE
.PP
Please report bugs in #tahoe\-lafs on Irc2p or via email (see below).
.SH SEE ALSO
.PP
The \f[I]grid\-updates\f[] Git repositories:
.IP \[bu] 2
http://darrob.i2p/grid\-updates/
.PD 0
.P
.PD
.IP \[bu] 2
http://git.repo.i2p/r/grid\-updates.git
.PD 0
.P
.PD
.IP \[bu] 2
http://killyourtv.i2p/git/grid\-updates.git
.PP
The README on the grid:
.IP
.nf
\f[C]
URI:DIR2\-RO:hgh5ylzzj6ey4a654ir2yxxblu:hzk3e5rbsefobeqhliytxpycop7ep6qlscmw4wzj5plicg3ilotq/README.html
\f[]
.fi
.PP
Information about Tahoe\-LAFS for I2P and the I2P grid:
http://killyourtv.i2p
.SH LICENSE
.PP
\f[I]grid\-updates\f[] has been released into the public domain.
This means that you can do whatever you please with it.
.SH AUTHORS
darrob <darrob@mail.i2p>, KillYourTV <killyourtv@mail.i2p>.


CHANGES
2014-06-04 (version 1.2.1)
        * Initial support for Tahoe-LAFS 1.10. Support for displaying news in an iframe is not yet
	  supported on this version of Tahoe.
	* Fix compatibility check when run on systems with Tahoe-LAFS 1.10
	* Automatically add `http://` to node URLS that lack them.
	* Refuse to run on Python 3.0 or 3.1
	* Addition of some basic unit tests
	* Other minor fixes and improvements

2012-09-09 (version 1.2.0)
	* Subscription URIs need to be file caps now
	* Changed repair routines
		- Removed former --repair; renamed --repair-list to --repair
		- Changed the default repair subscription to the former community repair list
	* Added a new repair mode (level-check)
	* Changed source file format for --make-news to reStructuredText
	* Removed arch as an option to --format
	* Switched from Optparse to Argparse
	* Catch non-JSON responses from Tahoe (see ticket #1799)
	* Minor bug fixes and improvements

2012-06-30 (version 1.1.3)
	* Add arch as an option for --format
	* Bugfixes:
		- make-news: links from news will now open outside of the IFRAME.
		- Don't assume __init__.py was compiled when looking for the tahoe
		  node directory.

2012-06-04 (version 1.1.2)
	* Support running grid-updates as user root if it can be determined user root
	  is running an instance of Tahoe.

2012-04-30 (version 1.1.1)

	* Support patching Tahoe-LAFS 1.8.3 to 1.9.3
	* Bugfixes:
		- --undo-patch-tahoe and --patch-tahoe can be run as user root
		- Only inform of updated patches if Tahoe was patched

2012-04-29 (version 1.1.0)

	* Renamed --community-repair to --repair-list
	* Renamed --comrepair-uri to --repairlist-uri
	* Changed expected filename community-repair.json.txt to
	  repair-list.json.txt
	* Improved JSON list format (incompatible to 1.0.0)
	* --download-update now takes an argument (tar, deb, zip, exe, py2exe)
	* Split gridupdates.py into separate modules
	* Changed default Tahoe locations of script updates and the introducer list
	* Various bug fixes

2012-04-20 (version 1.0.0)

	* This is a complete Python rewrite, which extends compatibility to BSD and
	  Windows.
	* changed some actions:
		- --replace-introducers/-r has been replaced by a better
		  --sync-introducers/-s action
		- '-r' is now --repair-shares
		- '-R' is now --community-repairs (see below)
	* new --community-repair action: this is an additional subscription that
	  lets you participate in the maintenance of the included shares.
	* new '-q' (quiet) option (the program is a little more verbose now; for
	  cron jobs you may want to use -q.)
	* new config file format (INI); see man page
	* introducer subscription list is now in JSON format, which breaks
	  compatibility with older versions
	* removed ability to download from other URIs than Tahoe's
		- removed --no-proxy option
		- removed --download-tool option
	* Minor: removed use of `markdown` again
	* Minor: removed --create-config action

2012-02-28 (version 0.7.3)

	* Use `markdown` if `pandoc` isn't available
	* Bugfixes: fail gracefully if markdown compilers or tahoe can't be
	  located

2012-02-10 (version 0.7.2)

	* Bugfix: only inform about updated patches if Web UI is already patched;
	  leave users with no interest in the patch alone

2012-02-09 (version 0.7.1)

	* Fixed a bug preventing the script to detect new patch versions
	* Updated Tahoe web UI patch and NEWS template to include links to the ATOM
	  news feed

2012-02-01 (version 0.7)

	* Fixed major bug which broke introducer list management
	* Made new NEWS be printed in whole, not a diff
	* Made code less I2P-specific
	* Various minor bug fixes and code improvements

2012-01-28 (version 0.6.2)

	* Various mostly minor bug fixes

2012-01-20 (version 0.6.1)

	* Bug fixes regarding the --patch-tahoe function

2012-01-20 (version 0.6)

	* Added actions to patch (and restore) the Tahoe web conole
	* Added --make-news action
	* Added uninstall target to Makefile
	* Added action to create a config file template
	* Various bug fixes and improvements

2012-01-12 (version 0.5)

	* Re-enabled eepget for I2P versions > 0.8.12
	* Support for configuration files
	* New --download-tool option
	* Include Atom news feed
	* Change default LIST URI to a commented introducers list
	* Various improvements and bugfixes (not yet BSD compatible again)

2012-01-04 (version 0.4.1)

	* Disable eepget until bug http://trac.i2p2.i2p/ticket/576 is fixed.
	* Another BSD bugfix
	* Various minor code and documentation clean-ups
	* Change default NEWS URI

2011-12-09 (version 0.4)

	* New news format: NEWS.tgz. The NEWS tarball contains the old plain text
	  NEWS as well as an html version which will be placed into the Tahoe-LAFS
	  web server's web.static directory.

2011-12-08 (version 0.3)

	* New installation methods (`make install` and deb package)
	* New --no-proxy option to bypass the HTTP proxy
	* Updated documentation, including a man page
	* Various minor improvements

2011-12-04 (version 0.2)

	* Various bug fixes (also checked for BSD compatibility)
	* Better documentation (README and INSTALL)
	* Much more flexible download options
	* Renamed some actions to hopefully clarify their purpose
	  (you may have to adapt your cron setups)
	* Renamed grid-updates.sh to grid-updates

2011-12-02 (version 0.1)

	First release


grid-updates.sh
#!/bin/sh

# grid-updates -- a helper script for a Tahoe-LAFS nodes
#
#
# Authors:
#     darrob <darrob@mail.i2p>
#     KillYourTV <kytv@mail.i2p>
#
# License: public domain
#
# Requirements:
#     Bourne-compatible shell, Tahoe-LAFS

# Introduction
# ============
# This script retrieves and installs an up-to-date list of introducers for
# Tahoe-LAFS nodes that support multiple introducers.  See
# http://killyourtv.i2p/tahoe-lafs/ for more information.
#
# Run this script with either --merge-introducers or --replace-introducers as a
# cron job to make sure your Tahoe-LAFS node will get the most reliable
# connection to the I2P grid as possible.
#
# The list is stored on the grid itself and -- like all other shares -- needs
# maintenance and repairs.  If you can, please also add the
# --repair-subscriptions action to your cron job, or run it separately every
# once in a while.  This is in everyone's interest.
#
# If you also want to receive news relevant to the grid, add the
# --download-news action.  It will fetch and display a NEWS file from the grid.
# This is recommended.

# Setup notes
# ===========
# For this script to work, it needs read and write permissions to your
# Tahoe-LAFS node's directory (typically ~/.tahoe).  It will update your
# introducers file (if you ask it to) and make a backup of it.  If you also
# fetch news, the script will write them to a file called NEWS.

# This script's version:
VERSION='0.1-git'

############################### Configuration #################################
# Default location of the Tahoe-LAFS node:
#TAHOE_NODE_DIR="$HOME/.tahoe"
# Default location (directory) of the subscription list:
LISTURI='URI:DIR2-RO:22s6zidugdxaeikq6lakbxbcci:mgrc3nfnygslyqrh7hds22usp6hbn3pulg5bu2puv6y3wpoaaqqq'
# Default location (directory) of the NEWS file:
NEWSURI='URI:DIR2-RO:vi2xzmrimvcyjdoypphdwxqbte:g7lpf2v6vyvl4w5udgpriiawg6ofmbazktvxmspesvkqtmujr2rq/Latest'
# Default location (directory) of script releases:
SCRIPTURI='URI:DIR2-RO:mjozenx3522pxtqyruekcx7mh4:eaqgy2gfsb73wb4f4z2csbjyoh7imwxn22g4qi332dgcvfyzg73a'
###############################################################################

only_verbose () {
	if [ $OPT_VERBOSE ]; then
		$@
	fi
}

# Stop multiple instances from running simultaneously
if [ -w /var/lock ]; then  # the default lock directory in Linux
        LOCKDIR="/var/lock/grid-updates.lck"
else
        LOCKDIR="/tmp/grid-updates.lck"   # but maybe not elsewhere...
fi
PIDFILE="${LOCKDIR}/PID"
ENO_SUCCESS=0;
ENO_GENERAL=1;
ENO_LOCKFAIL=2;
ENO_RECVSIG=3;

if mkdir "${LOCKDIR}"  > /dev/null 2>&1 ; then
       trap 'ECODE=$?;
       rm -rf "${LOCKDIR}"' 0
       touch $PIDFILE
       echo $$ > "${PIDFILE}"
       trap 'echo "ERROR: Killed by a signal $ECODE $ENO_RECVSIG" >&2
            exit ${ENO_RECVSIG}' 1 2 3 15
else
       # lock failed, check if it's stale
       OTHERPID="$(cat "${PIDFILE}")"

       if [ $? != 0 ]; then
               echo "ERROR: Another instance of `basename $0` is active with PID ${OTHERPID}" >&2
               exit ${ENO_LOCKFAIL}
       fi

       if ! kill -0 ${OTHERPID} >/dev/null 2>&1; then
               #stale lock, removing it and restarting
               only_verbose echo "INFO: Removing stale PID ${OTHERPID}" >&2
               rm -rf ${LOCKDIR} || echo "Cannot remove $LOCKDIR" >&2; exit 1
               only_verbose echo "INFO: [`basename $0`] restarting" >&2
               exec "$0" "$@"
       else
               #lock is valid and OTHERPID is active
               echo "ERROR: Another instance of `basename $0` is active with PID ${OTHERPID}" >&2
               exit ${ENO_LOCKFAIL}
       fi
fi

print_help () {
cat << EOF

Usage: $0 [OPTIONS] [ACTION]

Actions:
    -m, --merge-introducers     Merge your local introducers list with the
                                subscription's
    -r, --replace-introducers   Replace your local list of introducers with the
                                master list
    -n, --download-news         Retrieve news regarding the I2P grid.  These
                                will be stored in [node directory]/NEWS.
                                If you run this script as a cron job, the
                                news will also be emailed to you.
    --check-version             Check for a new version of this script on the
                                grid
    --download-update           Download a new version of this script from the
        [target directory]      grid (implies --check-version)
    -R, --repair-subscriptions  Maintain or repair the health of the subscription
                                service's URI
Options:
    -d [directory],             Specify the node directory (default: ~/.tahoe)
    --node-directory [directory]
    --list-uri [URI]            Override default location of introducers
                                list
    --news-uri [URI]            Override default location of news file
    --script-uri [URI]          Override default location of script updates
    -v, --verbose               Display more verbose output
    -V, --version               Display version information
    -h, --help                  Print this help text

Errors:
    If the script repeatedly fails to retrieve a file from the grid, the share
    may be damaged.  Try running --repair-subscriptions which will try to
    repair it.  If that doesn't help, you will most likely have to find a new
    URI to subscribe to.  Ask in #tahoe-lafs on Irc2P, check the DeepWiki
    and/or http://killyourtv.i2p.

EOF
}

TAHOE=$(which tahoe)
[ -z "$TAHOE" ] && echo "ERROR: \`tahoe\` executable not found." >&2 && exit 1

check_tahoe_node () {
	if [ -d $TAHOE_NODE_DIR ]; then
		if [ ! -e $TAHOE_NODE_DIR/node.url ]; then
			echo "ERROR: Cannot find $TAHOE_NODE_DIR/node.url. Not a valid tahoe node." >&2
			exit 1
		fi
		return 0
	else
		echo "ERROR: $TAHOE_NODE_DIR is not a directory." >&2
		exit 1
	fi
}

# Abort if any variables aren't initialized to try to prevent any surprises
set -o nounset  # same as set -u
set -e          # abort if there are any uncaught errors along the way

check_permissions () {
	if [ ! -w $TAHOE_NODE_DIR ]; then
		echo "ERROR: Need write permissions to "$TAHOE_NODE_DIR" to be able to update "$1"" >&2
		exit 1
	fi
	if [ -e "$2" ] && [ ! -w "$2" ]; then
		echo "ERROR: Need write permissions to \"$2\" to be able to update "$1"" >&2
		exit 1
	fi
}

pretty_print () {
	while read line ; do echo "$line" | sed 's/^/INFO:\ \ \ \ /'; done
}

download_list () {
	only_verbose echo "INFO: Attempting to download introducers list."
	TMPLIST=$(mktemp $LOCKDIR/grid-update.XXXX)
	if [ $(echo $LISTURI |egrep '^http:\/\/(.*).i2p') ]; then
		if [ ! $(which wget) ]; then
			echo "ERROR: wget is required for retrieving subscriptions from eepsites." >&2
			return 1
		fi
		if ! wget --quiet "$LISTURI" -O "$TMPLIST" >/dev/null 2>&1 ; then
			echo "ERROR: Could not retrieve the list from $LISTURI." >&2
			return 1
		fi
	else
		if ! "$TAHOE" get "$LISTURI"/introducers "$TMPLIST" 2> /dev/null ; then
			echo "ERROR: Could not retrieve the list. Try again or check the share's integrity. See \`$0 --help.\`" >&2
			return 1
		fi
	fi
}

backup_list () {
	if [ -e "$INTRODUCER_LIST" ]; then
		LISTBAK="$INTRODUCER_LIST.bak"
		if [ ! -w "$LISTBAK" ] && ! touch "$LISTBAK" 2> /dev/null ; then
			echo "ERROR: Need write permissions to $LISTBAK to be able to update the file." >&2
			exit 1
		fi
		echo "# This is a backup of "$INTRODUCER_LIST". It was created by `basename $0` on $(date -u)." > "$LISTBAK"
		cat "$INTRODUCER_LIST" >> "$LISTBAK"
		return 0
	fi
}

merge_list () {
	INTRODUCER_LIST="$TAHOE_NODE_DIR/introducers"
	TMPLIST2=$(mktemp $LOCKDIR/grid-update.XXXX)
	if [ ! -e "$INTRODUCER_LIST" ]; then
		only_verbose echo "INFO: Unable to find "$INTRODUCER_LIST". Retrieving a new list."
		replace_list
		return 0
	else
		# Add new URIs in the subscribed list to the local list.
		# This resembles I2P's address book's system.
		check_permissions "the subscription list" "$INTRODUCER_LIST"
		download_list
		backup_list
		cat "$INTRODUCER_LIST.bak" "$TMPLIST" \
			| grep '^pb://' | sort -u > "$TMPLIST2"
		if diff -N "$TMPLIST2" "$INTRODUCER_LIST" > /dev/null 2>&1 ; then
			only_verbose echo "INFO: No new introducers found."
			return 0
		else
			cat "$INTRODUCER_LIST.bak" "$TMPLIST" \
				| grep '^pb://' | sort -u > "$INTRODUCER_LIST"  # merge
			only_verbose echo "INFO: Success: the list has been retrieved and merged. (Changes will take effect upon restart of the node.)"
		fi

		return 0
	fi
}

replace_list () {
	INTRODUCER_LIST="$TAHOE_NODE_DIR/introducers"
	# Make the local list identical to the subscribed one.
	check_permissions "the subscription list" "$INTRODUCER_LIST"
	download_list
	backup_list
	if diff -N "$LISTBAK" "$INTRODUCER_LIST" > /dev/null 2>&1; then
		mv -f "$TMPLIST" "$INTRODUCER_LIST" > /dev/null 2>&1   # install list
		only_verbose echo "INFO: Success: the list has been retrieved. (Changes will take effect upon restart of the node.)"
	elif [ ! -e "$INTRODUCER_LIST" ]; then # if we're here there was a back up but no introducer list
		mv -f "$TMPLIST" "$INTRODUCER_LIST" > /dev/null 2>&1   # install list
		only_verbose echo "INFO: Success: the list has been retrieved. (Changes will take effect upon restart of the node.)"
	else
		only_verbose echo "INFO: No updates available."
	fi
	return 0
}

checking_failed ()
{
	echo "ERROR: failed to check $1 share." >&2
	return 1
}

deep_check () {
	if [ $OPT_VERBOSE ]; then
		"$TAHOE" deep-check -v --repair --add-lease "$1" 2>/dev/null >&1 | pretty_print
	else
		"$TAHOE" deep-check --repair --add-lease "$1" >/dev/null 2>&1
	fi
}

repair_subscriptions () {
	check_tahoe_node

	only_verbose echo "INFO: Beginning to check subscription shares."
	only_verbose echo "INFO: Checking subscription share (1/3)."
	deep_check "$LISTURI" || checking_failed "subscriptions"

	only_verbose echo "INFO: Checking NEWS share (2/3)."
	deep_check "$NEWSURI" || checking_failed "news"

	only_verbose echo "INFO: Checking scripts share (3/3)."
	deep_check "$SCRIPTURI" || checking_failed "scripts"
}

print_news () {
	# if new NEWS...
	if ! diff -N "$TAHOENEWS" "$TMPNEWS" > /dev/null ; then
		# move old news aside
		if [ -e "$TAHOENEWS" ]; then
			cp -f "$TAHOENEWS" "$OLDNEWS"
		fi
		# put new news into place
		cp -f "$TMPNEWS" "$TAHOENEWS" > /dev/null
		# print
		diff --ignore-all-space --ignore-blank-lines --new-file \
		        "$OLDNEWS" "$TAHOENEWS" | grep -e "^>\ " | sed 's/^>\ //'
		return 0
	else
		only_verbose echo "INFO: There are no news."
		return 0
	fi
}

download_news () {
	TAHOENEWS="$TAHOE_NODE_DIR/NEWS"
	check_permissions "news" "$TAHOENEWS"
	TMPNEWS=$(mktemp $LOCKDIR/grid-update.XXXX)
	OLDNEWS=$(mktemp $LOCKDIR/grid-update.XXXX)
	only_verbose echo "INFO: Attempting to download NEWS file."
	if [ $(echo $NEWSURI |egrep '^http:\/\/(.*).i2p') ]; then
		if [ ! $(which wget) ]; then
			echo "ERROR: wget is required for retrieving news from eepsites." >&2
			return 1
		fi
		if  wget --quiet "$NEWSURI" -O "$TMPNEWS" > /dev/null 2>&1 ; then
			print_news
			return 0
		else
			echo "ERROR: Could not retrieve the news from $NEWSURI." >&2
			return 1
		fi
	else
		if "$TAHOE" get "$NEWSURI/NEWS" "$TMPNEWS" 2> /dev/null ; then
			print_news
			return 0
		else
			echo "ERROR: Couldn't fetch the NEWS file." >&2
			exit 1
		fi
	fi
}

check_uri () {
	# Tahoe URIs will start with URI:
	#Yes, this is very rudimentary checking,
	# but it's better than nothing...isn't it?

	if [ ! $(echo "$1" | grep '^URI:') ] && [ ! $(echo "$1" | egrep '^http\:\/\/(.*).i2p') ] ; then
		echo "ERROR: "$1" is not a valid ${2}-uri." >&2
		return 1
	fi
}

check_version () {
	if [ $(echo $SCRIPTURI |egrep '^http:\/\/(.*).i2p') ]; then
		echo "ERROR: Script updates are not currently supported from eepSites." >&2
		return 1
	fi
	LC_ALL=C # make sure sort order is what we expect
	only_verbose echo "INFO: Attempting to check for new version."
	# sort -V not on FreeBSD. These sort parameters were taken from http://stackoverflow.com/a/4046149
	LATEST_VERSION_FILENAME=$(tahoe ls $SCRIPTURI | grep 'grid-updates-v[[:digit:]]\.[[:digit:]].*\.tgz' | sort -r -t- -k2,2 -n | head -n 1)
	if [ "$LATEST_VERSION_FILENAME" ]; then
		LATEST_AVAILABLE_VERSION=$(echo $LATEST_VERSION_FILENAME | sed 's/^grid-updates-v\(.*\)\.tgz$/\1/')
		NEWEST_VERSION=$(/bin/echo -e "$VERSION\n$LATEST_AVAILABLE_VERSION" | sort -r -t- -k2,2 -n | head -n 1)
		if [ "$VERSION" != "$NEWEST_VERSION" ]; then
			# Only print if not called via --download-update
			[ ! $OPT_DOWNLOAD_UPDATE ] && echo "New version available: $LATEST_AVAILABLE_VERSION."
			return 0
		else
			# Only print if not called via --download-update
			[ ! $OPT_DOWNLOAD_UPDATE ] && echo "No new version available."
			return 1
		fi
	else
		echo "Error: Nothing found. This doesn't seem to be a download location for grid-updates." >&2
	fi
}

download_update () {
	if [ ! -d $UPDATE_DOWNLOAD_DIR ]; then
		echo "ERROR: $UPDATE_DOWNLOAD_DIR is not a directory." >&2
		exit 1
	fi

	if [ ! -w $UPDATE_DOWNLOAD_DIR ]; then
		echo "ERROR: Cannot write to download directory $UPDATE_DOWNLOAD_DIR." >&2
		exit 1
	fi

	if check_version; then
		only_verbose echo "INFO: New version available: $LATEST_AVAILABLE_VERSION."
		only_verbose echo "INFO: Attempting to download new version."
		if tahoe get $SCRIPTURI/$LATEST_VERSION_FILENAME "$UPDATE_DOWNLOAD_DIR/$LATEST_VERSION_FILENAME" 2> /dev/null ; then
			echo "Update found (version $LATEST_AVAILABLE_VERSION) and downloaded to $UPDATE_DOWNLOAD_DIR/$LATEST_VERSION_FILENAME."
		fi
	else
		only_verbose echo "INFO: No new version available."
	fi
}

# Useful constants
: ${TAHOE_NODE_DIR:="$HOME/.tahoe"}
: ${http_proxy:="127.0.0.1:4444"}
export http_proxy

# some of those variables will not be initialized--and that's OK.
# We'll do our own checking from this point forward.
set +u

while [ $# -gt 0 ] ; do
	case $1 in
		--node-directory|-d)
			if [ -z "$2" ]; then
				echo "ERROR: Tahoe node directory not specified." >&2
				print_help
				exit 1
			fi
			TAHOE_NODE_DIR=$2
			shift 2
			check_tahoe_node
		;;
		--list-uri)
			if [ -z "$2" ]; then
				echo "ERROR: list-uri not specified." >&2
				print_help
				exit 1
			fi
			LISTURI=$2
			shift 2
			check_uri "$LISTURI" "list"
		;;
		--news-uri)
			if [ -z "$2" ]; then
				echo "ERROR: news-uri not specified." >&2
				print_help
				exit 1
			fi
			NEWSURI=$2
			shift 2
			check_uri "$NEWSURI" "news"
		;;
		--script-uri)
			if [ -z "$2" ]; then
				echo "ERROR: script-uri not specified." >&2
				print_help
				exit 1
			fi
			SCRIPTURI=$2
			shift 2
			check_uri "$SCRIPTURI" "script"
		;;
		--merge-introducers|-m)
			OPT_MERGE_LIST=1
			shift
		;;
		--replace-introducers|-r)
			OPT_REPLACE_LIST=1
			shift
		;;
		--repair-subscriptions|-R)
			OPT_REPAIR_SUBSCRIPTIONS=1
			shift
		;;
		--download-news|-n)
			OPT_DOWNLOAD_NEWS=1
			shift
		;;
		--version|-V)
			echo "$(basename $0) version: $VERSION"
			exit 0
		;;
		--check-version)
			OPT_CHECK_VERSION=1
			shift
		;;
		--download-update)
			if [ -z "$2" ]; then
				only_verbose echo "INFO: No download directory specified. Using $PWD/."
				UPDATE_DOWNLOAD_DIR="$PWD"
				OPT_DOWNLOAD_UPDATE=1
				shift
			else
				UPDATE_DOWNLOAD_DIR="$2"
				OPT_DOWNLOAD_UPDATE=1
				shift 2
			fi
		;;
		--verbose|-v)
			OPT_VERBOSE=1
			shift
		;;
		--help|-h)
			print_help
			exit
		;;
		*)
			echo "ERROR: Unknown command." >&2
			print_help
			exit 1
		;;
	esac
done

if [ ! $OPT_MERGE_LIST ] && [ ! $OPT_REPLACE_LIST ] && \
[ ! $OPT_REPAIR_SUBSCRIPTIONS ] && [ ! $OPT_DOWNLOAD_NEWS ] && \
[ ! $OPT_CHECK_VERSION ] && [ ! $OPT_DOWNLOAD_UPDATE ]; then
	echo "ERROR: You need to specify an action." >&2
	print_help
	exit 1
fi
[ $OPT_REPAIR_SUBSCRIPTIONS ] && repair_subscriptions
if [ $OPT_MERGE_LIST ] && [ ! $OPT_REPLACE_LIST ]; then
	merge_list
elif [ ! $OPT_MERGE_LIST ] && [ $OPT_REPLACE_LIST ]; then
	replace_list
elif [ $OPT_MERGE_LIST ] && [ $OPT_REPLACE_LIST ]; then
	echo "ERROR: --merge-introducers and --replace-introducers are mutually exclusive." >&2
	print_help
	exit 1
fi
[ $OPT_DOWNLOAD_NEWS ] && download_news
if [ $OPT_CHECK_VERSION ] && [ $OPT_DOWNLOAD_UPDATE ]; then
	download_update
elif [ $OPT_DOWNLOAD_UPDATE ]; then
	download_update
elif [ $OPT_CHECK_VERSION ]; then
	check_version
fi

exit 0
